File: 02_Delta-3.1.3-IL (1)/Python/Solutions/Includes/Common-Notebooks/Assertion-Utils.py
# Databricks notebook source

import pyspark
from typing import Callable, Any, Iterable, List, Set, Tuple

#############################################
# Deprecated test functions
#############################################

def dbTest(id, expected, result):
  import uuid
  
  if id: eventId = "Test-"+id 
  else: eventId = "Test-"+str(uuid.uuid1())

  evaluation = str(expected) == str(result)
  status = "passed" if evaluation else "failed"
  daLogger.logEvent(id, f"{eventId}\nDBTest Assertion\n{status}\n1")

  assert evaluation, f"{result} does not equal expected {expected}"
  
#############################################
# Test Suite classes
#############################################

# Test case
class TestCase(object):
  __slots__=('description', 'testFunction', 'id', 'dependsOn', 'escapeHTML', 'points')
  def __init__(self,
               description:str,
               testFunction:Callable[[], Any],
               id:str=None,
               dependsOn:Iterable[str]=[],
               escapeHTML:bool=False,
               points:int=1):
    
    self.description=description
    self.testFunction=testFunction
    self.id=id
    self.dependsOn=dependsOn
    self.escapeHTML=escapeHTML
    self.points=points

# Test result
class TestResult(object):
  __slots__ = ('test', 'skipped', 'debug', 'passed', 'status', 'points', 'exception', 'message')
  def __init__(self, test, skipped = False, debug = False):
    try:
      self.test = test
      self.skipped = skipped
      self.debug = debug
      if skipped:
        self.status = 'skipped'
        self.passed = False
        self.points = 0
      else:
        assert test.testFunction() != False, "Test returned false"
        self.status = "passed"
        self.passed = True
        self.points = self.test.points
      self.exception = None
      self.message = ""
    except Exception as e:
      self.status = "failed"
      self.passed = False
      self.points = 0
      self.exception = e
      self.message = repr(self.exception)
      if (debug and not isinstance(e, AssertionError)):
        raise e

# Decorator to lazy evaluate - used by TestSuite
def lazy_property(fn):
    '''Decorator that makes a property lazy-evaluated.
    '''
    attr_name = '_lazy_' + fn.__name__

    @property
    def _lazy_property(self):
        if not hasattr(self, attr_name):
            setattr(self, attr_name, fn(self))
        return getattr(self, attr_name)
    return _lazy_property

  
testResultsStyle = """
<style>
  table { text-align: left; border-collapse: collapse; margin: 1em; caption-side: bottom; font-family: Sans-Serif; font-size: 16px}
  caption { text-align: left; padding: 5px }
  th, td { border: 1px solid #ddd; padding: 5px }
  th { background-color: #ddd }
  .passed { background-color: #97d897 }
  .failed { background-color: #e2716c }
  .skipped { background-color: #f9d275 }
  .results .points { display: none }
  .results .message { display: none }
  .results .passed::before  { content: "Passed" }
  .results .failed::before  { content: "Failed" }
  .results .skipped::before { content: "Skipped" }
  .grade .passed  .message:empty::before { content:"Passed" }
  .grade .failed  .message:empty::before { content:"Failed" }
  .grade .skipped .message:empty::before { content:"Skipped" }
</style>
    """.strip()


class __TestResultsAggregator(object):
  testResults = dict()
  
  def update(self, result:TestResult):
    self.testResults[result.test.id] = result
    return result
  
  @lazy_property
  def score(self) -> int:
    return __builtins__.sum(map(lambda result: result.points, self.testResults.values()))
  
  @lazy_property
  def maxScore(self) -> int:
    return __builtins__.sum(map(lambda result: result.test.points, self.testResults.values()))

  @lazy_property
  def percentage(self) -> int:
    return 0 if self.maxScore == 0 else int(100.0 * self.score / self.maxScore)

  def displayResults(self):
    displayHTML(testResultsStyle + f"""
    <table class='results'>
      <tr><th colspan="2">Test Summary</th></tr>
      <tr><td>Number of Passing Tests</td><td style="text-align:right">{self.score}</td></tr>
      <tr><td>Number of Failing Tests</td><td style="text-align:right">{self.maxScore-self.score}</td></tr>
      <tr><td>Percentage Passed</td><td style="text-align:right">{self.percentage}%</td></tr>
    </table>
    """)
# Lazy-man's singleton
TestResultsAggregator = __TestResultsAggregator()


# Test suite class
class TestSuite(object):
  def __init__(self, initialTestCases: Iterable[TestCase] = None) -> None:
    self.ids = set()
    self.testCases = list()
    if initialTestCases:
      for tC in initialTestCases:
        self.addTest(tC)
        
  @lazy_property
  def testResults(self) -> List[TestResult]:
    return self.runTests()
  
  def runTests(self, debug=False) -> List[TestResult]:
    import re
    import uuid
    failedTests = set()
    testResults = list()

    for test in self.testCases:
      skip = any(testId in failedTests for testId in test.dependsOn)
      result = TestResult(test, skip, debug)

      if (not result.passed and test.id != None):
        failedTests.add(test.id)

      if result.test.id: eventId = "Test-"+result.test.id 
      elif result.test.description: eventId = "Test-"+re.sub("[^a-zA-Z0-9_]", "", result.test.description).upper()
      else: eventId = "Test-"+str(uuid.uuid1())
      message = f"{eventId}\n{result.test.description}\n{result.status}\n{result.points}"
      daLogger.logEvent(eventId, message)

      testResults.append(result)
      TestResultsAggregator.update(result)
    
    return testResults

  def _display(self, cssClass:str="results", debug=False) -> None:
    from html import escape
    testResults = self.testResults if not debug else self.runTests(debug=True)
    lines = []
    lines.append(testResultsStyle)
    lines.append("<table class='"+cssClass+"'>")
    lines.append("  <tr><th class='points'>Points</th><th class='test'>Test</th><th class='result'>Result</th></tr>")
    for result in testResults:
      resultHTML = "<td class='result "+result.status+"'><span class='message'>"+result.message+"</span></td>"
      descriptionHTML = escape(str(result.test.description)) if (result.test.escapeHTML) else str(result.test.description)
      lines.append("  <tr><td class='points'>"+str(result.points)+"</td><td class='test'>"+descriptionHTML+"</td>"+resultHTML+"</tr>")
    lines.append("  <caption class='points'>Score: "+str(self.score)+"</caption>")
    lines.append("</table>")
    html = "\n".join(lines)
    displayHTML(html)
  
  def displayResults(self) -> None:
    self._display("results")
  
  def grade(self) -> int:
    self._display("grade")
    return self.score
  
  def debug(self) -> None:
    self._display("grade", debug=True)
  
  @lazy_property
  def score(self) -> int:
    return __builtins__.sum(map(lambda result: result.points, self.testResults))
  
  @lazy_property
  def maxScore(self) -> int:
    return __builtins__.sum(map(lambda result: result.test.points, self.testResults))

  @lazy_property
  def percentage(self) -> int:
    return 0 if self.maxScore == 0 else int(100.0 * self.score / self.maxScore)

  def addTest(self, testCase: TestCase):
    if not testCase.id: raise ValueError("The test cases' id must be specified")
    if testCase.id in self.ids: raise ValueError(f"Duplicate test case id: {testCase.id}")
    self.testCases.append(testCase)
    self.ids.add(testCase.id)
    return self
  
  def test(self, id:str, description:str, testFunction:Callable[[], Any], points:int=1, dependsOn:Iterable[str]=[], escapeHTML:bool=False):
    testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return self.addTest(testCase)
  
  def testEquals(self, id:str, description:str, valueA, valueB, points:int=1, dependsOn:Iterable[str]=[], escapeHTML:bool=False):
    testFunction = lambda: valueA == valueB
    testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return self.addTest(testCase)
    
  def testFloats(self, id:str, description:str, valueA, valueB, tolerance=0.01, points:int=1, dependsOn:Iterable[str]=[], escapeHTML:bool=False):
    testFunction = lambda: compareFloats(valueA, valueB, tolerance)
    testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return self.addTest(testCase)

  def testRows(self, id:str, description:str, rowA: pyspark.sql.Row, rowB: pyspark.sql.Row, points:int=1, dependsOn:Iterable[str]=[], escapeHTML:bool=False):
    testFunction = lambda: compareRows(rowA, rowB)
    testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return self.addTest(testCase)
  
  def testDataFrames(self, id:str, description:str, dfA: pyspark.sql.DataFrame, dfB: pyspark.sql.DataFrame, testColumnOrder: bool, testNullable: bool, points:int=1, dependsOn:Iterable[str]=[], escapeHTML:bool=False):
    testFunction = lambda: compareDataFrames(dfA, dfB, testColumnOrder, testNullable)
    testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return self.addTest(testCase)
  
  def testSchemas(self, id:str, description:str, schemaA: pyspark.sql.types.StructType, schemaB: pyspark.sql.types.StructType, testColumnOrder: bool, testNullable: bool, points:int=1, dependsOn:Iterable[str]=[], escapeHTML:bool=False):
    testFunction = lambda: compareSchemas(schemaA, schemaB, testColumnOrder, testNullable)
    testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return self.addTest(testCase)
  
  def testContains(self, id:str, description:str, listOfValues, value, points:int=1, dependsOn:Iterable[str]=[], escapeHTML:bool=False):
    testFunction = lambda: value in listOfValues
    testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return self.addTest(testCase)

#############################################
# Test Suite utilities
#############################################


def getQueryString(df: pyspark.sql.DataFrame) -> str:
  # redirect sys.stdout to a buffer
  import sys, io
  stdout = sys.stdout
  sys.stdout = io.StringIO()

  # call module
  df.explain(extended=True)

  # get output and restore sys.stdout
  output = sys.stdout.getvalue()
  sys.stdout = stdout

  return output


#############################################
# Test Suite comparison functions
#############################################

def compareFloats(valueA, valueB, tolerance=0.01):
  # Usage: compareFloats(valueA, valueB) (uses default tolerance of 0.01)
  #        compareFloats(valueA, valueB, tolerance=0.001)
  from builtins import abs 
  
  try:
    if (valueA == None and valueB == None):
         return True
      
    else:
         return abs(float(valueA) - float(valueB)) <= tolerance 
      
  except:
    return False
  
  
def compareRows(rowA: pyspark.sql.Row, rowB: pyspark.sql.Row):
  # Usage: compareRows(rowA, rowB)
  # compares two Dictionaries
  
  if (rowA == None and rowB == None):
    return True
  
  elif (rowA == None or rowB == None):
    return False
  
  else: 
    return rowA.asDict() == rowB.asDict()


def compareDataFrames(dfA: pyspark.sql.DataFrame, dfB: pyspark.sql.DataFrame, testColumnOrder: bool, testNullable: bool):
  if dfA == None and dfB == None: return True
  if dfA == None or dfB == None: return False
  if compareSchemas(dfA.schema, dfB.schema, testColumnOrder, testNullable) == False: return False
  if dfA.count() != dfB.count(): return False

  rowsA = dfA.collect()
  rowsB = dfB.collect()

  for i in range(0, len(rowsA)):
    rowA = rowsA[i]
    rowB = rowsB[i]
    for column in dfA.columns:
      valueA = rowA[column]
      valueB = rowB[column]
      if (valueA != valueB): return False

  return True


def compareSchemas(schemaA: pyspark.sql.types.StructType, schemaB: pyspark.sql.types.StructType, testColumnOrder: bool, testNullable: bool): 
  
  from pyspark.sql.types import StructField
  
  if (schemaA == None and schemaB == None): return True
  if (schemaA == None or schemaB == None): return False
  
  schA = schemaA
  schB = schemaB

  if (testNullable == False):  
      schA = [StructField(s.name, s.dataType, True) for s in schemaA]
      schB = [StructField(s.name, s.dataType, True) for s in schemaB]

  if (testColumnOrder == True):
    return [schA] == [schB]
  else:
    return set(schA) == set(schB)
  
displayHTML("Initializing Databricks Academy's testing framework...")

File: 02_Delta-3.1.3-IL (1)/Scala/Solutions/Includes/Common-Notebooks/Assertion-Utils.scala
// Databricks notebook source

//*******************************************
// Deprecated test functions
//*******************************************

def dbTest[T](id: String, expected: T, result: => T): Unit = {
  val eventId = "Test-" + (if (id != null) id else java.util.UUID.randomUUID)
  val evaluation = (result == expected)
  val status = if (evaluation) "passed" else "failed"
  daLogger.logEvent(id, s"$eventId\nDBTest Assertion\n$status\n1")

  assert(evaluation, s"$result does not equal $expected expected")
}

//*******************************************
// TEST SUITE CLASSES
//*******************************************

// Test case
case class TestCase(description:String,
                    testFunction:()=>Any,
                    id:String=null,
                    dependsOn:Seq[String]=Nil,
                    escapeHTML:Boolean=false,
                    points:Int=1)

// Test result
case class TestResult(test: TestCase, skipped:Boolean = false, debug:Boolean = false) {
  val exception: Option[Throwable] = {
    if (skipped)
      None
    else if (debug) {
      if (test.testFunction() != false)
        None
      else 
        Some(new AssertionError("Test returned false"))
    } else {
      try {
        assert(test.testFunction() != false, "Test returned false")
        None
      } catch {
        case e: Exception => Some(e)
        case e: AssertionError => Some(e)
      }
    }
  }

  val passed: Boolean = !skipped && exception.isEmpty

  val message: String = {
    exception.map(ex => {
      val msg = ex.getMessage()
      if (msg == null || msg.isEmpty()) ex.toString() else msg
    }).getOrElse("")
  }
  
  val status: String = if (skipped) "skipped" else if (passed) "passed" else "failed"
  
  val points: Int = if (passed) test.points else 0
}

val testResultsStyle = """
<style>
  table { text-align: left; border-collapse: collapse; margin: 1em; caption-side: bottom; font-family: Sans-Serif; font-size: 16px}
  caption { text-align: left; padding: 5px }
  th, td { border: 1px solid #ddd; padding: 5px }
  th { background-color: #ddd }
  .passed { background-color: #97d897 }
  .failed { background-color: #e2716c }
  .skipped { background-color: #f9d275 }
  .results .points { display: none }
  .results .message { display: none }
  .results .passed::before  { content: "Passed" }
  .results .failed::before  { content: "Failed" }
  .results .skipped::before { content: "Skipped" }
  .grade .passed  .message:empty::before { content:"Passed" }
  .grade .failed  .message:empty::before { content:"Failed" }
  .grade .skipped .message:empty::before { content:"Skipped" }
</style>
    """.trim

object TestResultsAggregator {
  val testResults = scala.collection.mutable.Map[String,TestResult]()
  def update(result:TestResult):TestResult = {
    testResults.put(result.test.id, result)
    return result
  }
  def score = testResults.values.map(_.points).sum
  def maxScore = testResults.values.map(_.test.points).sum
  def percentage = (if (maxScore == 0) 0 else 100.0 * score / maxScore).toInt
  def displayResults():Unit = {
    displayHTML(testResultsStyle + s"""
    <table class='results'>
      <tr><th colspan="2">Test Summary</th></tr>
      <tr><td>Number of Passing Tests</td><td style="text-align:right">${score}</td></tr>
      <tr><td>Number of Failing Tests</td><td style="text-align:right">${maxScore-score}</td></tr>
      <tr><td>Percentage Passed</td><td style="text-align:right">${percentage}%</td></tr>
    </table>
    """)
  }
}

// Test Suite
case class TestSuite(initialTestCases : TestCase*) {
  
  val testCases : scala.collection.mutable.ArrayBuffer[TestCase] = scala.collection.mutable.ArrayBuffer[TestCase]()
  val ids : scala.collection.mutable.ArrayBuffer[String] = scala.collection.mutable.ArrayBuffer[String]()
  
  initialTestCases.foreach(addTest)
  
  import scala.collection.mutable.ListBuffer
  import scala.xml.Utility.escape
  
  def runTests(debug:Boolean = false):scala.collection.mutable.ArrayBuffer[TestResult] = {
    val failedTests = scala.collection.mutable.Set[String]()
    val results = scala.collection.mutable.ArrayBuffer[TestResult]()
    for (testCase <- testCases) {
      val skip = testCase.dependsOn.exists(failedTests contains _)
      val result = TestResult(testCase, skip, debug)

      if (result.passed == false && testCase.id != null) {
        failedTests += testCase.id
      }
      val eventId = "Test-" + (if (result.test.id != null) result.test.id 
                          else if (result.test.description != null) result.test.description.replaceAll("[^a-zA-Z0-9_]", "").toUpperCase() 
                          else java.util.UUID.randomUUID)
      val message = s"${eventId}\n${result.test.description}\n${result.status}\n${result.points}"
      daLogger.logEvent(eventId, message)
      
      results += result
      TestResultsAggregator.update(result)
    }
    return results
  }

  lazy val testResults = runTests()

  private def display(cssClass:String="results", debug:Boolean=false) : Unit = {
    val testResults = if (!debug) this.testResults else runTests(debug=true)
    val lines = ListBuffer[String]()
    lines += testResultsStyle
    lines += s"<table class='$cssClass'>"
    lines += "  <tr><th class='points'>Points</th><th class='test'>Test</th><th class='result'>Result</th></tr>"
    for (result <- testResults) {
      val resultHTML = s"<td class='result ${result.status}'><span class='message'>${result.message}</span></td>"
      val descriptionHTML = if (result.test.escapeHTML) escape(result.test.description) else result.test.description
      lines += s"  <tr><td class='points'>${result.points}</td><td class='test'>$descriptionHTML</td>$resultHTML</tr>"
    }
    lines += s"  <caption class='points'>Score: $score</caption>"
    lines += "</table>"
    val html = lines.mkString("\n")
    displayHTML(html)
  }
  
  def displayResults() : Unit = {
    display("results")
  }
  
  def grade() : Int = {
    display("grade")
    score
  }
  
  def debug() : Unit = {
    display("grade", debug=true)
  }
  
  def score = testResults.map(_.points).sum
  
  def maxScore : Integer = testCases.map(_.points).sum

  def percentage = (if (maxScore == 0) 0 else 100.0 * score / maxScore).toInt
  
  def addTest(testCase: TestCase):TestSuite = {
    if (testCase.id == null) throw new IllegalArgumentException("The test cases' id must be specified")
    if (ids.contains(testCase.id)) throw new IllegalArgumentException(f"Duplicate test case id: {testCase.id}")
    testCases += testCase
    ids += testCase.id
    return this
  }
  
  def test(id:String, description:String, testFunction:()=>Any, points:Int=1, dependsOn:Seq[String]=Nil, escapeHTML:Boolean=false): TestSuite = {
    val testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return addTest(testCase)
  }
  
  def testEquals(id:String, description:String, valueA:Any, valueB:Any, points:Int=1, dependsOn:Seq[String]=Nil, escapeHTML:Boolean=false): TestSuite = {
    val testFunction = ()=> (valueA == valueB)
    val testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return addTest(testCase)
  }
  
  def testFloats(id:String, description:String, valueA: Any, valueB: Any, tolerance: Double=0.01, points:Int=1, dependsOn:Seq[String]=Nil, escapeHTML:Boolean=false): TestSuite = {
    val testFunction = ()=> compareFloats(valueA, valueB, tolerance)
    val testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return addTest(testCase)
  }
  
  def testRows(id:String, description:String, rowA: org.apache.spark.sql.Row, rowB: org.apache.spark.sql.Row, points:Int=1, dependsOn:Seq[String]=Nil, escapeHTML:Boolean=false): TestSuite = {
    val testFunction = ()=> compareRows(rowA, rowB)
    val testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return addTest(testCase)
  }
  
  def testDataFrames(id:String, description:String, dfA:org.apache.spark.sql.DataFrame, dfB:org.apache.spark.sql.DataFrame, testColumnOrder:Boolean, testNullable:Boolean, points:Int=1, dependsOn:Seq[String]=Nil, escapeHTML:Boolean=false): TestSuite = {
    val testFunction = ()=> compareDataFrames(dfA, dfB, testColumnOrder, testNullable)
    val testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return addTest(testCase)
  }
  
  def testSchemas(id:String, description:String, schemaA: org.apache.spark.sql.types.StructType, schemaB: org.apache.spark.sql.types.StructType, testColumnOrder: Boolean, testNullable: Boolean, points:Int=1, dependsOn:Seq[String]=Nil, escapeHTML:Boolean=false): TestSuite = {
    val testFunction = ()=> compareSchemas(schemaA, schemaB, testColumnOrder, testNullable)
    val testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return addTest(testCase)
  }

  def testContains(id:String, description:String, list:Seq[Any], value:Any, points:Int=1, dependsOn:Seq[String]=Nil, escapeHTML:Boolean=false): TestSuite = {
    val testFunction = ()=> list.contains(value)
    val testCase = TestCase(id=id, description=description, testFunction=testFunction, dependsOn=dependsOn, escapeHTML=escapeHTML, points=points)
    return addTest(testCase)
  }
}

//*******************************************
// SCORE FUNCTIONS AND PLACEHOLDER GENERATION
//*******************************************

def getQueryString(df: org.apache.spark.sql.Dataset[Row]): String = {
  df.explain.toString
}

//*******************************************
// Test Suite comparison functions
//*******************************************

def compareFloats(valueA: Any, valueB: Any, tolerance: Double=0.01): Boolean = {
  // Usage: compareFloats(valueA, valueB) (uses default tolerance of 0.01)
  //        compareFloats(valueA, valueB, tolerance=0.001)
  
  import scala.math
  try{
       if (valueA == null && valueB == null) {
         true
       } else if (valueA == null || valueB == null) {
         false
       } else {
         math.abs(valueA.asInstanceOf[Number].doubleValue - valueB.asInstanceOf[Number].doubleValue) <= tolerance 
       }
  } catch {
    case e: ClassCastException => {
      false
   }   
  }
}


def compareRows(rowA: org.apache.spark.sql.Row, rowB: org.apache.spark.sql.Row): Boolean = {
  // Usage: compareRows(rowA, rowB)
  // compares two rows as unordered Maps

  if (rowA == null && rowB == null) {
    true
    
  } else if (rowA == null || rowB == null) {
    false
    
  // for some reason, the schema didn't make it
  } else if (rowA.schema == null || rowB.schema == null) {
    rowA.toSeq.toSet == rowB.toSeq.toSet
    
  } else {
    rowA.getValuesMap[String](rowA.schema.fieldNames.toSeq) == rowB.getValuesMap[String](rowB.schema.fieldNames.toSeq)
  }
}


def compareDataFrames(dfA: org.apache.spark.sql.DataFrame, dfB: org.apache.spark.sql.DataFrame, testColumnOrder: Boolean, testNullable: Boolean): Boolean = {
  if (dfA == null && dfB == null) return true
  if (dfA == null || dfB == null) return false
  if (compareSchemas(dfA.schema, dfB.schema, testColumnOrder, testNullable) == false) return false
  if (dfA.count != dfB.count) return false

  val rowsA = dfA.collect()
  val rowsB = dfB.collect()
  
  for (i <- 0 until rowsA.length) {
    val rowA = rowsA(i)
    val rowB = rowsB(i)
    for (column <- dfA.columns) {
      val valueA = rowA.getAs[Any](column)
      val valueB = rowB.getAs[Any](column)
      if (valueA != valueB) return false
    }
  }
  return true
}


def compareSchemas(schemaA: org.apache.spark.sql.types.StructType, schemaB: org.apache.spark.sql.types.StructType, testColumnOrder: Boolean, testNullable: Boolean): Boolean = {
  
  if (schemaA == null && schemaB == null) return true
  if (schemaA == null || schemaB == null) return false
  
  var schA = schemaA.toSeq
  var schB = schemaB.toSeq

  if (testNullable == false) {   
    schA = schemaA.map(_.copy(nullable=true)) 
    schB = schemaB.map(_.copy(nullable=true)) 
  }

  if (testColumnOrder == true) {
    schA == schB
  } else {
    schA.toSet == schB.toSet
  }
}

displayHTML("Initializing Databricks Academy's testing framework...")

